src/coord/rpnet.c

```diff
@@ -336,12 +557,22 @@ int rpnetNext(ResultProcessor *self, SearchResult *r) {
   if (new_reply) {
     if (resp3) { // RESP3
       nc->curIdx = 0;
-      nc->base.parent->totalResults += MRReply_Length(rows);
+      // Note: For WITHCOUNT in multi-shard aggregate, totalResults is already set
+      // by the waiting logic above. We skip accumulation here to avoid double-counting.
+      // For non-WITHCOUNT or single-shard cases, we still need to count.
+      if (!HasWithCount(nc->areq) || !(nc->areq->reqflags & QEXEC_F_IS_AGGREGATE)) {
```

Comment:
This check should not be in the execution path. Can we move it to the iterator constructor? Or at least check using some other hint, don't rely on the hard-coded check copied from the parser